DOC: Clarify docs on mixed advanced indexing and slicing

paulrougieux opened the pull request:
It was not clear how to reproduce the slice example with multi dimensional array indexing. I hope the additional example clarifies the equivalence between slice and multi array for new users.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

paulrougieux committed:
Complement slice example with it's equivalent as multi dimentional array index

It was not clear how to reproduce the slice example with multi dimensional array indexing. I hope the additional example clarifies the equivalence between slice and multi array for new users.

eric-wieser commented on numpy/doc/indexing.py:
```diff
@@ -294,6 +294,13 @@
 np.array([[1,2]]) (shape (1,2)) that is broadcast with the index array
 to produce a resultant array of shape (3,2).
```
Hmm, this sentence is wrong too - this isn't broadcasting at all, as `(3,)` and `(1, 2)` do not broadcast.

I wonder if we should instead explain this as `y[:,1:3][np.array([0,2,4]),:]`? @seberg, thoughts?

seberg replied:
Uhh, right. I guess saying that it is the same as slicing everything first, and then taking the whole slice in the second index would work.
A bit ugly, but at least it makes the notion clear that there is in fact *no* broadcasting, etc. going on and slices always use orthogonal/outer semantics (they work independently).

So yeah, I think that is the best solution I can think of. Saying that they work independently and thus be separated out like that.

anirudh2290 replied:
> I wonder if we should instead explain this as `y[:,1:3][np.array([0,2,4]),:]`?
good point! maybe it is obvious, but it would be nice to put it in two python statements to indicate slicing followed by the index array usage.

anirudh2290 replied:
To push this forward: based on the above suggestions, I suggested a change here : https://github.com/numpy/numpy/pull/15891#discussion_r404341840 . I wasn't able to suggest change on L293, so we can commit this one and remove the L293 in a following commit. Alternatively, the code can just be pushed to @paulrougieux repo, but that probably needs to be done by a maintainer.

eric-wieser commented on numpy/doc/indexing.py:
```diff
@@ -294,6 +294,13 @@
 np.array([[1,2]]) (shape (1,2)) that is broadcast with the index array
 to produce a resultant array of shape (3,2).
 
+This is equivalent to: ::
+
+ >>> y[np.array([[0],[2],[4]]),np.array([[1,2]])]
+ array([[ 1,  2],
+        [15, 16],
+        [29, 30]])
```
Style nits:
```suggestion
This is equivalent to::

 >>> y[np.array([[0], [2], [4]]), np.array([[1, 2]])]
 array([[ 1,  2],
        [15, 16],
        [29, 30]])
```

anirudh2290 commented on numpy/doc/indexing.py:
```diff
@@ -294,6 +294,13 @@
 np.array([[1,2]]) (shape (1,2)) that is broadcast with the index array
 to produce a resultant array of shape (3,2).
 
+This is equivalent to: ::
+
+ >>> y[np.array([[0],[2],[4]]),np.array([[1,2]])]
+ array([[ 1,  2],
+        [15, 16],
+        [29, 30]])
```
```suggestion
In effect, the slice and index array operation are independent.
The slice operation extracts the 1st and 3rd columns in y,
followed by the index array operation which extracts the 0th, 2nd
and 4th rows from the sliced array.

This is equivalent to: ::

>>> y[:, 1:3][np.array([0, 2, 4]), :]
```

seberg replied:
Thanks, I applied your changes manually, since they span outside of what github can do with the suggest API...
@paulrougieux you used your "master" branch here, which means that these commits will have been pushed to your master branch. Once this is all finished, you will want to forceably reset your master to the NumPy master. Generally, sooner or later it is always painful when working directly on "master" for pull requests.

anirudh2290 commented:
gentle ping @seberg, @eric-wieser. alternatively, @paulrougieux it would be great if you can push this change if possible.

seberg committed:
Update numpy/doc/indexing.py

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

seberg committed:
DOC,fixup: Manually apply suggestions from @anirudh2290

paulrougieux commented:
There is an issue with the column name comment:
`The slice operation extracts the 1st and 3rd columns in y,`
should be `The slice operation extracts the 1st and 2nd columns in y,`



Trying to figure out how to update this pull request. I didn't use git on my laptop in this simple case. I just edited directly in github.

paulrougieux committed:
Correct description of columns impacted by slice

English row and column description with ordinal numbers is more natural when index starts with one.

anirudh2290 commented:
> There is an issue with the column name comment:
The slice operation extracts the 1st and 3rd columns in y,
should be The slice operation extracts the 2nd and 3rd columns in y, i.e. the columns with index 1 and 2.

Thanks for pointing this out, nice catch !

> I realize the description of row numbers would then be inconsistent with the one for columns. You actually described the first row as 0th row. Is there a standard for describing row order with ordinal numbers? Zeroth row feels weird as an English description, but it obviously makes sense in the python indexing context. I think I'll put the ordinal numbers in parenthesis and simply leave the cardinal indexes in the text.

The way you did it looks good to me, I see other docs in numpy for eg. absolute beginners doc also using this approach.

anirudh2290 commented on numpy/doc/indexing.py:
```diff
@@ -285,14 +285,23 @@
+The slice operation extracts columns with index 1 and 2,
+(i.e. the 2nd and 3rd columns),
+followed by the index array operation which extracts rows with 
+index 0, 2 and 4 (i.e the first, 3rd and 5th rows).
```
nit:
```suggestion
index 0, 2 and 4 (i.e the first, third and fifth rows).
```
otherwise looks good to me.

seberg committed:
Update numpy/doc/indexing.py

Co-Authored-By: Anirudh Subramanian <anirudh2290@apache.org>

seberg commented:
Thanks for checking again @paulrougieux and Anirudh and Eric! Lets put this in then :), one unclear thing less in the documentation is always good.
